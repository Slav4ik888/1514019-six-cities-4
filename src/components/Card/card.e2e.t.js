import React from "react";
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Card from "./card.jsx";
import {testOffer} from "../../mocks/test-offer.js";
import configureStore from 'redux-mock-store';
import {NameSpace} from '../../reducers/name-space.js';
import {Provider} from 'react-redux';


const mockStore = configureStore([]);

Enzyme.configure({
  adapter: new Adapter(),
});

describe(`<Card> tests`, () => {

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const onCardTitleClick = jest.fn();
  const onCardFocusEnter = jest.fn();
  const onCardFocusLeave = jest.fn();
  const onFavClick = jest.fn();
  const isFav = false;

  const renderComponent = (props = {}) => {
    const store = mockStore({
      [NameSpace.USER]: {
        userStatus: `AUTH`,
      },
    });
    return shallow(
        <Provider store={store}>
          <Card offer = {testOffer}
            onCardTitleClick = {onCardTitleClick}
            onCardFocusEnter = {onCardFocusEnter}
            onCardFocusLeave = {onCardFocusLeave}
            onFavClick = {onFavClick}
            isFav = {isFav}
            {...props}
          />
        </Provider>
    );
  };

  it(`Клик по title`, () => {
    const placeComponent = renderComponent();
    const placeTitle = placeComponent.find(`h2.place-card__name a`);

    placeTitle.simulate(`click`);
    expect(onCardTitleClick).toHaveBeenCalledTimes(1);
  });

  it(`Навели мышку на карточку`, () => {
    const component = renderComponent();
    const card = component.find(`article.place-card`);

    card.simulate(`pointerenter`);
    expect(onCardFocusEnter).toHaveBeenCalledTimes(1);
  });

  it(`Убрали мышку с карточки`, () => {
    const component = renderComponent();
    const card = component.find(`article.place-card`);

    card.simulate(`pointerleave`);
    expect(onCardFocusLeave).toHaveBeenCalledTimes(1);
  });

  it(`Кликнули по флажку в избранное`, () => {
    const component = renderComponent();
    const card = component.find(`place-card__bookmark-button`);

    card.simulate(`click`);
    expect(onFavClick).toHaveBeenCalledTimes(1);
  });

});
