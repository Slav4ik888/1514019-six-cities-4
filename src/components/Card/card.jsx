import React from 'react';
import {Link} from 'react-router-dom';

import pt from 'prop-types';
import {offerPropTypes} from '../../utils/prop-types-templates.js';

import ButtonFavorite from '../ButtonFavorite/button-favorite.jsx';

import {getRating} from '../../utils/utils.js';
import {AppRoute, placesType} from '../../utils/const.js';


const Card = ({offer,
  // onCardTitleClick, isFav
  onCardFocusEnter, onCardFocusLeave,
  // onFavClick,
  type}) => {

  const {id, isPremium, previewImage, price, rating, cardTitle, offerType} = offer;

  // let favClass = isFav ? `place-card__bookmark-button--active` : null;

  // const handleTitleClick = () => {
  //   onCardTitleClick(offer);
  // };

  const handlePointerEnter = () => {
    if (onCardFocusEnter) {
      onCardFocusEnter(offer);
    }
  };

  const handlePointerLeave = () => {
    if (onCardFocusLeave) {
      onCardFocusLeave();
    }
  };

  // const handleFavClick = () => {
  //   if (type !== placesType.FAVORITE) {
  //     onFavClick(offer);
  //   }
  // };

  let placeCard;
  let imageWrapper;
  let placeCardInfo;
  let imgWidth;
  let imgHeight;

  switch (type) {
    case placesType.CITY:
      placeCard = `cities__place-card place-card`;
      imageWrapper = `cities__image-wrapper place-card__image-wrapper`;
      placeCardInfo = `place-card__info`;
      imgWidth = 260; imgHeight = 200;
      break;

    case placesType.NEAR:
      placeCard = `near-places__card place-card`;
      imageWrapper = `near-places__image-wrapper place-card__image-wrapper`;
      placeCardInfo = `place-card__info`;
      imgWidth = 260; imgHeight = 200;
      break;

    case placesType.FAVORITE:
      placeCard = `favorites__card place-card`;
      imageWrapper = `favorites__image-wrapper place-card__image-wrapper`;
      placeCardInfo = `favorites__card-info place-card__info`;
      // favClass = `place-card__bookmark-button--active`;
      imgWidth = 150; imgHeight = 110;
      break;
  }

  return (
    <article className={placeCard}
      onPointerEnter={handlePointerEnter}
      onPointerLeave={handlePointerLeave}
    >
      {isPremium &&
        <div className="place-card__mark">
          <span>Premium</span>
        </div>}

      <div className={imageWrapper}>
        <a href="#">
          <img className="place-card__image" src={previewImage} width={imgWidth} height={imgHeight} alt="Place image" />
        </a>
      </div>
      <div className={placeCardInfo}>
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <ButtonFavorite
            offer={offer}
            type={type}
          />

          {/* <button className={`place-card__bookmark-button button ${favClass}`} type="button"
            onClick={handleFavClick}
          >
            <svg className="place-card__bookmark-icon" width="18" height="19">
              <use xlinkHref="#icon-bookmark"></use>
            </svg>
            <span className="visually-hidden">To bookmarks</span>
          </button> */}
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: getRating(rating)}}></span>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <Link className="place-card__name_href"
            to={`${AppRoute.ROOM}/${id}`}
          >
            {cardTitle}
          </Link>
        </h2>
        <p className="place-card__type">{offerType}</p>
      </div>
    </article>
  );
};

Card.propTypes = {
  onCardTitleClick: pt.func.isRequired,
  onCardFocusEnter: pt.func,
  onCardFocusLeave: pt.func,
  offer: pt.shape(offerPropTypes).isRequired,
  // isFav: pt.bool,
  // onFavClick: pt.func,
  type: pt.oneOf([placesType.CITY, placesType.NEAR, placesType.FAVORITE, placesType.OFFER_DETAILS]).isRequired,

};


export default Card;
