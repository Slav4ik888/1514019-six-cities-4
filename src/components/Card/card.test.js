import React from 'react';
import renderer from 'react-test-renderer';
import Card from './card.jsx';

const offerTitles = [
  `Beautiful & luxurious apartment at great location`,
  `Wood and stone place`,
  `Beautiful & luxurious apartment at great location`,
  `Wood and stone place`, `Wood and stone place`,
  `Beautiful & luxurious apartment at great location`,
  `Wood and stone place`
];

const cardTitleHandler = () => {};

describe(`<Main /> рендерит <Card>`, () => {
  it(`<Card> со offers ${offerTitles[0]}`, () => {
    const tree = renderer
      .create(<Card
        offerTitle={offerTitles[0]}
        onCardTitleClick={cardTitleHandler}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});
