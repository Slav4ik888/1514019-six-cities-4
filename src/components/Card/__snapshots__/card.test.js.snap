// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot <Card> Render <Card> type='CITY' isFav=false 1`] = `
<article
  className="cities__place-card place-card"
  onPointerEnter={[Function]}
  onPointerLeave={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Place image"
        className="place-card__image"
        height={200}
        src="img/apartment-03.jpg"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          180
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button null"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        className="place-card__name_href"
        href="/offer"
        onClick={[Function]}
      >
        Beautiful & luxurious apartment at great location
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      House
    </p>
  </div>
</article>
`;

exports[`Snapshot <Card> Render <Card> type='CITY' isFav=true 1`] = `
<article
  className="cities__place-card place-card"
  onPointerEnter={[Function]}
  onPointerLeave={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Place image"
        className="place-card__image"
        height={200}
        src="img/apartment-03.jpg"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          180
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button place-card__bookmark-button--active"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        className="place-card__name_href"
        href="/offer"
        onClick={[Function]}
      >
        Beautiful & luxurious apartment at great location
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      House
    </p>
  </div>
</article>
`;

exports[`Snapshot <Card> Render <Card> type='NEAR' isFav=false 1`] = `
<article
  className="near-places__card place-card"
  onPointerEnter={[Function]}
  onPointerLeave={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="near-places__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Place image"
        className="place-card__image"
        height={200}
        src="img/apartment-03.jpg"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          180
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button null"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        className="place-card__name_href"
        href="/offer"
        onClick={[Function]}
      >
        Beautiful & luxurious apartment at great location
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      House
    </p>
  </div>
</article>
`;

exports[`Snapshot <Card> Render <Card> type='NEAR' isFav=true 1`] = `
<article
  className="near-places__card place-card"
  onPointerEnter={[Function]}
  onPointerLeave={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="near-places__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Place image"
        className="place-card__image"
        height={200}
        src="img/apartment-03.jpg"
        width={260}
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          180
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button place-card__bookmark-button--active"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        className="place-card__name_href"
        href="/offer"
        onClick={[Function]}
      >
        Beautiful & luxurious apartment at great location
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      House
    </p>
  </div>
</article>
`;
